specs/experimental/alligator.md

```diff
@@ -0,0 +1,444 @@
+
+### Getters
+
+For backwards compatibility, the `Alligator` MUST implement all public getter functions of the
+`GovernanceToken` related to delegation and voting power. These functions MUST be used by the
+`GovernanceToken` when an account has been migrated to the `Alligator` contract. Otherwise,
+the `GovernanceToken` MUST use its own state. Similarly, all of the `Alligator` getter functions
+MUST use the `GovernanceToken` state if the account has not been migrated.
+
+#### `checkpoints`
+
+Retrieves the checkpoints for a given user address.
+
+```solidity
+function checkpoints(address _account) external view returns (ERC20Votes.Checkpoint[] memory)
+```
+
+#### `numCheckpoints`
+
+Retrieves the number checkpoints for a given user address.
+
+```solidity
+function numCheckpoints(address _account) external view returns (uint32)
+```
+
+#### `delegates`
+
+Retrieves the delegations of a given user address sorted in descending order by voting power.
+This function is intended to be used by the `GovernanceToken` contract to maximize for backwards compatibility.
+
+```solidity
+function delegates(address _account) external view returns (Delegation[] memory)
+```
+
+#### `getVotes`
+
+Retrieves the voting power of a given user address.
+
+```solidity
+function getVotes(address _account) external view returns (uint256)
+```
+
+#### `getPastVotes`
+
+Retrieves the voting power of a given user address at a block.
+
+```solidity
+function getPastVotes(address _account, uint256 _blockNumber) external view returns (uint256)
+```
+
+#### `getPastTotalSupply`
+
+Retrieves the total supply of the `GovernanceToken` at a given block.
+
+```solidity
+function getPastTotalSupply(uint256 _blockNumber) external view returns (uint256)
+```
+
+#### `migrated`
+
+Returns the migration status of an account — `True` if the account has been migrated, `False` otherwise.
+
+```solidity
+function migrated(address _account) public view returns (bool)
+```
+
+### Events
+
+#### `DelegationCreated`
+
+MUST trigger when an account delegates voting power to another address (delegatee).
+
+```solidity
+event DelegationCreated(address indexed account, Delegation delegation);
+```
+
+#### `DelegationsCreated`
+
+MUST trigger when an account delegates voting power to multiple addresses (delegatees).
+
+```solidity
+event DelegationsCreated(address indexed account, Delegation[] delegations);
+```
+
+#### `DelegateVotesChanged`
+
+MUST trigger every time the voting power of a user changes, including when a token transfer occurs or when a
+delegation is created.
+
+```solidity
+event DelegateVotesChanged(address indexed delegatee, uint256 previousBalance, uint256 newBalance)
+```
+
+## Storage
+
+The `Alligator` contract MUST be able to store delegations and checkpoints. These storage variables MUST be
+defined as in the `GovernanceToken` and use the same types:
+
+```solidity
+// The maximum number of delegations allowed.
+uint256 public constant MAX_DELEGATIONS = 100;
```

Comment:
These would be good to go in the constants table where the token address is defined

Reply:
we are getting the token address from the Predeploys lib, so should we still have a constants section in these specs with the `MAX_DELEGATIONS` and `DENOMINATOR` constants?

Reply:
Yeah we typically reserve a section for constants with a table like

| Constant Name | Value |
| --------------- | ----- |
| `MY_CONSTANT` | 123 |

Reply:
added in https://github.com/ethereum-optimism/specs/pull/247/commits/fef1220fb763c6f0e25b2d71f5f76ba21391445f